framework/src/modules/token/cc_method.ts

```diff
@@ -13,191 +13,270 @@
+	public init(ownChainID: Buffer): void {
+		this._ownChainID = ownChainID;
 	}
 
 	public addDependencies(interoperabilityMethod: InteroperabilityMethod) {
 		this._interopMethod = interoperabilityMethod;
 	}
 
-	public async beforeApplyCCM(_ctx: BeforeApplyCCMsgMethodContext): Promise<void> {
-		// const { ccm } = ctx;
-		// const methodContext = ctx.getMethodContext();
-		// if (ccm.fee < BigInt(0)) {
-		// 	throw new Error('Fee must be greater or equal to zero.');
-		// }
-		// const { id: ownChainID } = await this._interopMethod.getOwnChainAccount(methodContext);
-		// const { messageFeeTokenID } = await this._interopMethod.getChannel(
-		// 	methodContext,
-		// 	ccm.sendingChainID,
-		// );
-		// const [feeTokenChainID, feeTokenLocalID] = splitTokenID(messageFeeTokenID);
-		// const userStore = this.stores.get(UserStore);
-		// if (!feeTokenChainID.equals(ownChainID)) {
-		// 	await userStore.addAvailableBalanceWithCreate(
-		// 		methodContext,
-		// 		ctx.trsSender,
-		// 		messageFeeTokenID,
-		// 		ccm.fee,
-		// 	);
-		// 	return;
-		// }
-		// const escrowStore = this.stores.get(EscrowStore);
-		// await escrowStore.deductEscrowAmountWithTerminate(
-		// 	methodContext,
-		// 	this._interopMethod,
-		// 	ccm.sendingChainID,
-		// 	feeTokenLocalID,
-		// 	ccm.fee,
-		// );
-		// const canonicalTokenID = await this._tokenMethod.getCanonicalTokenID(
-		// 	methodContext,
-		// 	messageFeeTokenID,
-		// );
-		// await userStore.addAvailableBalanceWithCreate(
-		// 	methodContext,
-		// 	ctx.trsSender,
-		// 	canonicalTokenID,
-		// 	ccm.fee,
-		// );
+	public async beforeCrossChainCommandExecution(ctx: BeforeApplyCCMsgMethodContext): Promise<void> {
+		const { trsSender, ccm } = ctx;
+		const methodContext = ctx.getMethodContext();
+		const relayerAddress = crypto.address.getAddressFromPublicKey(trsSender);
+		const tokenID = await this._interopMethod.getMessageFeeTokenID(
+			methodContext,
+			ccm.sendingChainID,
+		);
+		const [chainID] = splitTokenID(tokenID);
+		const userStore = this.stores.get(UserStore);
+
+		if (chainID.equals(this._ownChainID)) {
+			const escrowStore = this.stores.get(EscrowStore);
+			const escrowKey = escrowStore.getKey(ccm.sendingChainID, tokenID);
+			const escrowAccount = await escrowStore.get(methodContext, escrowKey);
+
+			if (escrowAccount.amount < ccm.fee) {
+				this.events.get(BeforeCCCExecutionEvent).error(
+					methodContext,
+					{
+						sendingChainID: ccm.sendingChainID,
+						receivingChainID: ccm.receivingChainID,
+						messageFeeTokenID: tokenID,
+						messageFee: ccm.fee,
+						relayerAddress,
+					},
+					TokenEventResult.FAIL_INSUFFICIENT_BALANCE,
+				);
+
+				throw new Error('Insufficient balance in the sending chain for the message fee.');
+			}
+
+			escrowAccount.amount -= ccm.fee;
+			await escrowStore.set(methodContext, escrowKey, escrowAccount);
+		}
+
+		await userStore.addAvailableBalanceWithCreate(methodContext, relayerAddress, tokenID, ccm.fee);
```

Comment:
I think with the update, you can use `addAvailableBalance` in general because account always needs to be initialized separately now, so it's ok to fail if the account doesn't exist. Same for below